### ISSUE TITLE
Support `bun run --workspace` 

### ISSUE THREAD

**Issue by swissspidy** [oven-sh/bun#2232]
### What version of Bun is running?

0.5.7

### What platform is your computer?

Darwin 22.3.0 arm64 arm

### What steps can reproduce the bug?

In a monorepo with npm workspaces, scripts defined in a specific package's `package.json` can be run with 

```sh
npm run my-script --workspace packages/my-package
```

### What is the expected behavior?

Bun should allow me to run the package's script via the `--workspace` arg.

A possible workaround is:

```sh
(cd packages/my-package && bun run my-script)
```

### What do you see instead?

error: missing script "my-script"

**Comment by Electroid** [comment]
Thanks for reporting, we were talking about implementing this the other day. We'll do it

**Comment by EdwardIrby** [comment]
Don't know if this is still an issue but I was able to but I was able to get the following working

### What version of Bun is running?
0.5.9

### What platform is your computer?
WSL Unbuntu 

### What steps can reproduce the bug?
In a monorepo with npm workspaces, scripts defined in a specific package's package.json can be run with

`npm run my-script --workspace packages/my-package`

### What is the expected behavior?
Bun should allow me to run the package's script via the --workspace arg

### Additional information
Works as expected

**Comment by ferdy-roz** [comment]
Is this planned for the near future? this might be the one thing preventing us from migrating from pnpm to bun as without it we would have to do the `cd` workaround @swissspidy mentioned, which makes `package.json` very very chunky.

Really I want support for the `--workspaces` flag (plural), or some equivalent of pnpm's `-r`

**Comment by itpropro** [comment]
For bun as a drop-in replacement, it would really help to have `--workspace(s)` `--filter` and other monorepo related tooling available.
Most things with bun work extremely well and much better than I expected, are there any plans to improve the monorepo support in the near future @Electroid ?

**Comment by sheijne** [comment]
> (cd packages/my-package && bun run my-script)

As a an alternative workaround, `--cwd` seems to work: `bun run --cwd packages/my-package build`.

> Really I want support for the --workspaces flag (plural), or some equivalent of pnpm's -r

Would be nice to be able to reference a workspace based on the package name, ie: `bun run --workspace @my-org/my-package build`.

On top of this it'd be great to have wildcards: `bun run --workspace @my-org/my-*`

**Comment by kyle-belle** [comment]
any updates on this?

**Comment by xlc** [comment]
It will be great if `bun run` can support this feature from [`yarn run`](https://yarnpkg.com/cli/run#details)

> Otherwise, if the specified name contains a colon character and if one of the workspaces in the project contains exactly one script with a matching name, then this script will get executed.

**Comment by hyoretsu** [comment]
Bump, it's a bummer that such a great tool doesn't have this feature.

**Comment by Blankeos** [comment]
+1 on this. Would be useful to be able to do something like this. I'm actually currently using the `--cwd` that @sheijne mentioned. Thanks!

### What I'm doing
There's a **Root package.json** and an **apps/server/package.json (a Workspace)**.
- In the **root package.json**, I only do `bun run --cwd` on scripts from apps/server/package.json.
- In **apps/server/package.json**, I write the actual script I wanna execute.

```json
{
  // ./package.json (In Root of monorepo) | NOTICE HOW IT JUST RUNS PACKAGE.JSON SCRIPTS WITH BUN
  "scripts": {
    "server:dev": "bun run --cwd apps/server dev",
    "db:start": "bun run --cwd apps/server db:start",
    "db:stop": "bun run --cwd apps/server db:stop",
    "migrate:status": "bun run --cwd apps/server migrate:status",
    "migrate:create": "bun run --cwd apps/server migrate:create",
    "migrate:up": "bun run --cwd apps/server migrate:up",
    "migrate:down": "bun run --cwd apps/server migrate:down"
  },
}
```

```json
{
  // apps/server/package.json
  "name": "server",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "dev": "bun run --watch src/index.ts",
    "db:start": "docker-compose up -d --build",
    "db:stop": "docker-compose stop",
    "migrate:create": "sh scripts/migrate-create.sh src/database/migrations",
    "migrate:status": "dotenv -- goose -dir src/database/migrations postgres $DATABASE_URL status",
    "migrate:up": "dotenv -- goose -dir src/database/migrations postgres $DATABASE_URL up",
    "migrate:down": "dotenv -- goose -dir src/database/migrations postgres $DATABASE_URL down",
    "seed": "bun run scripts/seed.ts"
  }
}
```

### 1 thing to watch out for:
When you have migration scripts that use `dotenv-cli` the `.env` file of the folder, the above example doesn't work for 3 scripts there:

- ✅ `migrate:create` Works 
- ❎ `migrate:status`, `migrate:up`, and `migrate:down` DO NOT WORK. 

Why? I'm actually not completely sure why. But it does come up when I use the "root" version of the script. It works properly when I use the "apps/server" (workspace) version of the script. I almost ruled it out as a Bun workspace issue.

But actually the **culprit is [dotenv-cli](https://www.npmjs.com/package/dotenv-cli)**.

#### Solution for this
1. "Subscript Encapsulation". (Fortunately they have docs for this).
2. Another is just encapsulating the in an `.sh` file and let dotenv-cli run that script. Whcih turns out, I was already doing it for `migrate:create` (This is something I figured out accidentally on my own)

```json
    "migrate:create": "sh scripts/migrate-create.sh src/database/migrations",
    - "migrate:status": "dotenv -- goose -dir src/database/migrations postgres $DATABASE_URL status",
    - "migrate:up": "dotenv -- goose -dir src/database/migrations postgres $DATABASE_URL up",
    - "migrate:down": "dotenv -- goose -dir src/database/migrations postgres $DATABASE_URL down",
```
Changed to:
```json
    "migrate:create": "sh scripts/migrate-create.sh src/database/migrations",
    + "_migrate:status": "goose -dir src/database/migrations postgres $DATABASE_URL status",
    + "migrate:status": "dotenv -- bun run _migrate:status",
    + "_migrate:up": "goose -dir src/database/migrations postgres $DATABASE_URL up",
    + "migrate:up": "dotenv -- bun run _migrate:up",
    + "_migrate:down": "goose -dir src/database/migrations postgres $DATABASE_URL down",
    "migrate:down": "dotenv -- bun run _migrate:down",
```

I'll just leave these comments here for anyone attempting to do the same thing with Bun.

**Comment by amritk** [comment]
Are there any current workarounds to run multiple commands at once?

**Comment by choephix** [comment]
Is there complexity involved in shipping this feature out?
Being able to easily run a specific workspace's script is the only reason we're not switching to **bun** yet, tbh

**Comment by rhuanbarreto** [comment]
For now we are using [moonrepo](https://moonrepo.dev/docs/guides/javascript/bun-handbook) with full bun integration to be the task runner.

**Comment by zemse** [comment]
I'd like to propose additional helpers for bun workspaces:

- `bun add --workspace`: useful to install package in a workspace package from while in root wd itself
- `bun run --workspaces --if-present`: execute a script in all workspaces if it is present. E.g. I want to build everything.

**Comment by wesleycoder** [comment]
For those like me stumbling into this nowadays, I'm following [this recommendation](https://github.com/oven-sh/bun/issues/5350#issuecomment-1722382676) by @cs-clarence on #5350 to use [Thinkmill/manypkg](https://github.com/Thinkmill/manypkg) by adding it and this script to my root `package.json`:
```json
{
  "name": "my-project",
  "private": true,
  "scripts": {
    "pkg": "manypkg run"
  },
  "workspaces": [
    "apps/*",
    "packages/*"
  ],
  "devDependencies": {
    "@manypkg/cli": "^0.21.3"
  }
}
```
With this we can just `bun pkg <children-pattern> <script>` ex:
`bun pkg @proj dev` -> should start all `dev` scripts on sub-packages containing `@proj` on their name
See more on [manypkg docs](https://github.com/Thinkmill/manypkg#usage)

**Comment by Jarred-Sumner** [comment]
Bun v1.1.4 adds support for `bun --filter` which is a more powerful version of this feature. You can run workspace scripts in parallel (following dependency order). It works whether or not you're using `"workspaces"` in package.json and without `bun install`. 

The equivalent of `npm run --workspace` looks something like:
```sh
bun --filter="my-workspace-pkg" my-script-name
```

To run the `build` script in all the workspace packages that have`util` in the name:

```sh
bun run --filter="*util*" build
```

We'll have a page for this in bun.sh/docs later today

![gif](https://github.com/oven-sh/bun/assets/709451/0278e320-68c7-4506-9f83-cfe499ba9f18)

Fixed in https://github.com/oven-sh/bun/pull/8185

**Comment by johnnybenson** [comment]
Better or not, it's a bit unhelpful and comes off a touch arrogant to introduce a new api before delivering parity with an existing one that is well established and nearly everyone already understands and relies upon. I don't mean to be negative. Thanks for making filter. Filter is probably dope.

**Comment by hyoretsu** [comment]
Bro, it's literally a drop-in replacement for running commands in a workspace, if you want to do just that you can. But it also has so much more potential.
`bun --filter backend dev` and `bun --filter frontend dev` There you go. But you can also simply do `bun --filter "*" dev` and voila.

**Comment by wesleycoder** [comment]
@johnnybenson this is not a new api, anyone that used workspaces with pnpm will be familiar with `--filter`.
https://pnpm.io/filtering

**Comment by jylin** [comment]
The exit code for this features seems to always be 0. Should it exit 1 if any of the commands returns an error?

**Comment by wesleycoder** [comment]
I've tried to test this today using canary (`bun upgrade --canary`, which installed `1.1.4`)
But every usage I've tried always return `error: No packages matched the filter`

I've took a look to the [docs](https://github.com/oven-sh/bun/blob/main/docs/cli/filter.md) and the [`cli.zig`](https://github.com/oven-sh/bun/blob/fdaa01287aff7190a8b428d187186c4faa6d69bd/src/cli.zig#L190) on the main branch and unless I misunderstood I believed this was supposed to be on canary already.

Could anyone confirm if this is already available on canary?
If it was supposed to be working on canary I think it may be somewhat broken.
I'm able to provide more info or tests, just ping if needed :)

**Comment by dror-weiss** [comment]
@wesleycoder , I'm on 1.1.4 as well, and this message (`error: No packages matched the filter`) happened to me whenever there is a problem with the format. 

The filter feature supports calling **a script within the workspace**. I needed to execute a command in a workspace but it's not supported yet so I opened a new [issue](https://github.com/oven-sh/bun/issues/10321).

**Comment by wesleycoder** [comment]
I've tested with a Github Codespaces machine on [this minimal repository](https://github.com/wesleycoder/bun-test)

Neither of these works:
```sh
curl -fsSL https://bun.sh/install | bash
source /home/codespace/.bashrc
bun --version # 1.1.4

bun run --filter 'a' test
# error: No packages matched the filter

bun --filter 'a' test
# >>> Installed to the root package instead of installing to `packages/a`

bun upgrade --canary
bun --version # 1.1.4

bun run --filter 'a' test # same results
bun --filter 'a' test # same results
```

Now I believe it's not an issue on my machine and it's definitely broken, so I've open another issue: #10367

**Comment by kannthu** [comment]
Hey!

I tried switching from **Yarn** to **Bun** and I really missed the syntax of: `yarn workspace pkg build`

I don't really like the filter syntax - it is more annoying for me + It does not work with `add` command: `bun --filter pkg add imported_foo`

**So I wrote a script that brings back the `workspace` syntax to bun.**

### What it brings to bun:
```
bun workspace pkg build
bun workspace pkg add some_import
bun workspace pkg build --filters args_here
``` 

### How to do it:

**1. Add script to your root `package.json`:**
```
  "scripts": {
    "workspace": "./workspace.js",
   ...
  }
```

**2. Create `workspace.js` in the root directory of your project:**
```
#! /usr/bin/env node

const fs = require("fs");
const { execSync } = require("child_process");

const CACHE_FILE = "/tmp/vscode-bun-workspace-cache";
const CACHE_MAX_AGE = 300; // 5 minutes in seconds

function getCachedOutput() {
  try {
    // Check if cache exists and is fresh
    const stats = fs.statSync(CACHE_FILE);
    const age = (Date.now() - stats.mtimeMs) / 1000;

    if (age < CACHE_MAX_AGE) {
      return fs.readFileSync(CACHE_FILE, "utf-8");
    }
  } catch (err) {
    // Cache doesn't exist or other error
  }

  // Cache miss or stale - run command and cache result
  const output = execSync("bun turbo ls --output json").toString();
  fs.writeFileSync(CACHE_FILE, output);
  return output;
}

function main() {
  const args = process.argv.slice(2);

  // Exit if no package name provided
  if (args.length === 0) {
    console.error(
      "Please provide a package name - e.g. `bun workspace my-package`"
    );
    process.exit(1);
  }

  // Check for --no-cache flag
  let noCache = false;
  if (args[0] === "--no-cache") {
    noCache = true;
    args.shift();
  }

  const packageName = args[0];
  const output = noCache
    ? execSync("bun turbo ls --output json").toString()
    : getCachedOutput();

  const packages = JSON.parse(output).packages.items;
  const pkg = packages.find((p) => p.name === packageName);

  if (!pkg) {
    console.error(`Package '${packageName}' not found`);
    process.exit(1);
  }

  // Execute bun command with remaining arguments
  const remainingArgs = args.slice(1);
  const cmd =
    remainingArgs[0] === "add"
      ? `bun add --cwd "${pkg.path}" ${remainingArgs.slice(1).join(" ")}`
      : `bun --filter "./${pkg.path}" ${remainingArgs.join(" ")}`;

  try {
    execSync(cmd, { stdio: "inherit" });
  } catch (err) {
    process.exit(err.status);
  }
}

main();
```

**Comment by apuatcfbd** [comment]
> Bro, it's literally a drop-in replacement for running commands in a workspace, if you want to do just that you can. But it also has so much more potential. `bun --filter backend dev` and `bun --filter frontend dev` There you go. But you can also simply do `bun --filter "*" dev` and voila.

Is there way to run selected packages? Like I've package `A-D` want to run `A,B & D`
